toolchain/semantics/semantics_context.h

```diff
@@ -262,6 +265,27 @@ class SemanticsContext {
   auto ImplicitAsImpl(SemanticsNodeId value_id, SemanticsTypeId as_type_id,
                       SemanticsNodeId* output_value_id) -> ImplicitAsKind;
 
+  // Forms a canonical type ID for a type. This function is given two
+  // callbacks:
+  //
+  // `profile_type(id)` is called to determine the `FoldingSetNodeID id` for
+  // this type. The ID should be distinct for all distinct type values with the
+  // same `kind`.
+  //
+  // `make_node()` is called to obtain a `SemanticsNodeId` that describes the
+  // type. It is only called if the type does not already exist, so can be used
+  // to lazily build the `SemanticsNode`.
+  template <typename ProfileType, typename MakeNode>
+  auto CanonicalizeTypeImpl(SemanticsNodeKind kind, ProfileType profile_type,
+                            MakeNode make_node) -> SemanticsTypeId;
```

Comment:
What was the reason for the template approach for these callbacks? Can a std::function provide some structure?

Reply:
I'd suggest a std::function or function_ref, partly because I'm not clear the overhead is an issue. I'd rather go a template (with its type-checking and readability implications) only if a function has too much of a performance impact (and I also wonder if benchmarks might push in a different direction too for code structure)

Reply:
Done.